src/app/epics/user-login.epic.spec.ts

```diff
@@ -0,0 +1,176 @@
+                mockRouter as any
+            );
+        });
+
+        it('should initiate login and succeed', (done: DoneFn) => {
+            const user: string = 'USER';
+            const pass: string = 'PASSWORD';
+            const date: number = Date.now();
+
+            spyOn(mockLoginService, 'login')
+                .and
+                .returnValue( Observable.of(mockUser) );
+            spyOn(mockUserActions, 'loginSuccess')
+                .and
+                .returnValue( mockUserActions.loginSuccess(user, date) );
+            spyOn(mockUserActions, 'loginFailed')
+                .and
+                .returnValue( mockUserActions.loginFailed('Mock Error') )
+
+            action$ = ActionsObservable.of({ 
+                type: UserActions.LOGIN_USER_REQUEST,
+                username: user,
+                pass: pass,
+            });
+
+            userLoginEpic.login(action$)
+                .subscribe(
+                    () => {},
+
+                    () => done.fail('User Login Epic: login fail'),
+
+                    () => { 
+                        expect( mockLoginService.login )
+                            .toHaveBeenCalledWith(user, pass);
+                        expect( mockUserActions.loginSuccess )
+                            .toHaveBeenCalledWith(user, date);
+                        expect( mockUserActions.loginFailed )
+                            .not.toHaveBeenCalled();
+                        done();
+                    }
+                );
+        });
+        it('should throw error and failed action when login fails', (done: DoneFn) => {
+            const user: string = 'cheese';
+            const pass: string = 'cheese';
+            const date: number = Date.now();
+            const mockErrorMsg: string = 'MOCK_ERROR';
+
+            spyOn(mockLoginService, 'login')
+                .and
+                .returnValue( Observable.throw(mockErrorMsg) );
+            spyOn(mockUserActions, 'loginSuccess')
+                .and
+                .returnValue( mockUserActions.loginSuccess(user, date) );
+            spyOn(mockUserActions, 'loginFailed')
+                .and
+                .returnValue( mockUserActions.loginFailed(mockErrorMsg) );
+
+            action$ = ActionsObservable.of({ 
+                type: UserActions.LOGIN_USER_REQUEST,
+                username: user,
+                pass: pass,
+            });
+
+            userLoginEpic.login(action$)
+            .subscribe(
+                () => {},
+
+                () => done.fail('User Login Epic: login fail'),
+
+                () => { 
+                    expect( mockLoginService.login )
+                        .toHaveBeenCalledWith(user, pass);
+                    expect( mockUserActions.loginFailed )
+                        .toHaveBeenCalledWith(mockErrorMsg);
+                    expect( mockUserActions.loginSuccess )
+                        .not.toHaveBeenCalled();
+                    done();
```

Comment:
Similar guidance as above, rather than testing that the correct sequence of code has been invoked, it's better to test the meaningful outcome of having that code invoked - i.e. an API has been requested, if it failed, we get back the expected action